[JSC] Wasm GC BlockSignature should hold underlying TypeDefinition

Constellation opened the pull request:
#### 57538110b768d4e03c1779de5b7fc9e1cc1a4425
<pre>
[JSC] Wasm GC BlockSignature should hold underlying TypeDefinition
<a href="https://bugs.webkit.org/show_bug.cgi?id=282363">https://bugs.webkit.org/show_bug.cgi?id=282363</a>
<a href="https://rdar.apple.com/138177632">rdar://138177632</a>

Reviewed by Justin Michaud.

Wasm GC BlockSignature is not holding adhocly created TypeDefinition as
RefPtr. That&apos;s completely wrong: not holding TypeDefinition is allowed
because it is generated in wasm types in the section, and held by
ModuleInformation. If it is adhocly generated, it must be retained.

This patch fixes it by changing BlockSignature to keep underlying TypeDefinition
via RefPtr.

* JSTests/wasm/stress/block-signature-create-adhoc-type-definition.js: Added.
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::ControlData::branchTargetArity const):
(JSC::Wasm::BBQJITImpl::ControlData::branchTargetType const):
(JSC::Wasm::BBQJITImpl::ControlData::argumentType const):
(JSC::Wasm::BBQJITImpl::BBQJIT::addBlock):
(JSC::Wasm::BBQJITImpl::BBQJIT::addLoop):
(JSC::Wasm::BBQJITImpl::BBQJIT::addIf):
(JSC::Wasm::BBQJITImpl::BBQJIT::addElse):
(JSC::Wasm::BBQJITImpl::BBQJIT::addElseToUnreachable):
(JSC::Wasm::BBQJITImpl::BBQJIT::addTry):
(JSC::Wasm::BBQJITImpl::BBQJIT::addTryTable):
(JSC::Wasm::BBQJITImpl::BBQJIT::addReturn):
(JSC::Wasm::BBQJITImpl::BBQJIT::addEndToUnreachable):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedIfCompare):
* Source/JavaScriptCore/wasm/WasmFormat.h:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::splitStack):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseBlockSignatureAndNotifySIMDUseIfNeeded):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseBody):
(JSC::Wasm::FunctionParser&lt;Context&gt;::binaryCompareCase):
(JSC::Wasm::FunctionParser&lt;Context&gt;::unaryCompareCase):
(JSC::Wasm::FunctionParser&lt;Context&gt;::checkBranchTarget):
(JSC::Wasm::FunctionParser&lt;Context&gt;::unify):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseNestedBlocksEagerly):
(JSC::Wasm::FunctionParser&lt;Context&gt;::parseExpression):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntControlType::branchTargetType const):
(JSC::Wasm::IPIntControlType::branchTargetArity const):
(JSC::Wasm::IPIntGenerator::addElseToUnreachable):
(JSC::Wasm::IPIntGenerator::addEndToUnreachable):
(JSC::Wasm::IPIntGenerator::endTopLevel):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::ControlType::loop):
(JSC::Wasm::LLIntGenerator::ControlType::block):
(JSC::Wasm::LLIntGenerator::ControlType::if_):
(JSC::Wasm::LLIntGenerator::ControlType::createTry):
(JSC::Wasm::LLIntGenerator::ControlType::createTryTable):
(JSC::Wasm::LLIntGenerator::ControlType::branchTargetArity const):
(JSC::Wasm::LLIntGenerator::ControlType::branchTargetType const):
(JSC::Wasm::LLIntGenerator::addElseToUnreachable):
(JSC::Wasm::LLIntGenerator::addReturn):
(JSC::Wasm::LLIntGenerator::addEndToUnreachable):
(JSC::Wasm::LLIntGenerator::endTopLevel):
* Source/JavaScriptCore/wasm/WasmModuleInformation.cpp:
(JSC::Wasm::ModuleInformation::operator delete):
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
(JSC::Wasm::ModuleInformation::create): Deleted.
(JSC::Wasm::ModuleInformation::functionIndexSpaceSize const): Deleted.
(JSC::Wasm::ModuleInformation::isImportedFunctionFromFunctionIndexSpace const): Deleted.
(JSC::Wasm::ModuleInformation::typeIndexFromFunctionIndexSpace const): Deleted.
(JSC::Wasm::ModuleInformation::exceptionIndexSpaceSize const): Deleted.
(JSC::Wasm::ModuleInformation::isImportedExceptionFromExceptionIndexSpace const): Deleted.
(JSC::Wasm::ModuleInformation::typeIndexFromExceptionIndexSpace const): Deleted.
(JSC::Wasm::ModuleInformation::importFunctionCount const): Deleted.
(JSC::Wasm::ModuleInformation::internalFunctionCount const): Deleted.
(JSC::Wasm::ModuleInformation::importExceptionCount const): Deleted.
(JSC::Wasm::ModuleInformation::internalExceptionCount const): Deleted.
(JSC::Wasm::ModuleInformation::toCodeIndex const): Deleted.
(JSC::Wasm::ModuleInformation::toSpaceIndex const): Deleted.
(JSC::Wasm::ModuleInformation::memoryCount const): Deleted.
(JSC::Wasm::ModuleInformation::tableCount const): Deleted.
(JSC::Wasm::ModuleInformation::elementCount const): Deleted.
(JSC::Wasm::ModuleInformation::globalCount const): Deleted.
(JSC::Wasm::ModuleInformation::dataSegmentsCount const): Deleted.
(JSC::Wasm::ModuleInformation::table const): Deleted.
(JSC::Wasm::ModuleInformation::global const): Deleted.
(JSC::Wasm::ModuleInformation::initializeFunctionTrackers const): Deleted.
(JSC::Wasm::ModuleInformation::referencedFunctions const): Deleted.
(JSC::Wasm::ModuleInformation::hasReferencedFunction const): Deleted.
(JSC::Wasm::ModuleInformation::addReferencedFunction const): Deleted.
(JSC::Wasm::ModuleInformation::isDeclaredFunction const): Deleted.
(JSC::Wasm::ModuleInformation::addDeclaredFunction): Deleted.
(JSC::Wasm::ModuleInformation::isDeclaredException const): Deleted.
(JSC::Wasm::ModuleInformation::addDeclaredException): Deleted.
(JSC::Wasm::ModuleInformation::functionWasmSizeImportSpace const): Deleted.
(JSC::Wasm::ModuleInformation::functionWasmSize const): Deleted.
(JSC::Wasm::ModuleInformation::usesSIMDImportSpace const): Deleted.
(JSC::Wasm::ModuleInformation::usesSIMD const): Deleted.
(JSC::Wasm::ModuleInformation::markUsesSIMD): Deleted.
(JSC::Wasm::ModuleInformation::usesExceptions const): Deleted.
(JSC::Wasm::ModuleInformation::markUsesExceptions): Deleted.
(JSC::Wasm::ModuleInformation::usesAtomics const): Deleted.
(JSC::Wasm::ModuleInformation::markUsesAtomics): Deleted.
(JSC::Wasm::ModuleInformation::doneSeeingFunction): Deleted.
(JSC::Wasm::ModuleInformation::typeCount const): Deleted.
(JSC::Wasm::ModuleInformation::hasMemoryImport const): Deleted.
(JSC::Wasm::ModuleInformation::getBranchHint const): Deleted.
(JSC::Wasm::ModuleInformation::clobberingTailCalls const): Deleted.
(JSC::Wasm::ModuleInformation::callCanClobberInstance const): Deleted.
(JSC::Wasm::ModuleInformation::addClobberingTailCall): Deleted.
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::ControlData::ControlData):
(JSC::Wasm::OMGIRGenerator::ControlData::hasNonVoidresult const):
(JSC::Wasm::OMGIRGenerator::ControlData::branchTargetArity const):
(JSC::Wasm::OMGIRGenerator::ControlData::branchTargetType const):
(JSC::Wasm::OMGIRGenerator::addLoop):
(JSC::Wasm::OMGIRGenerator::addTopLevel):
(JSC::Wasm::OMGIRGenerator::addBlock):
(JSC::Wasm::OMGIRGenerator::addElseToUnreachable):
(JSC::Wasm::OMGIRGenerator::addEndToUnreachable):
* Source/JavaScriptCore/wasm/WasmParser.h:
(JSC::Wasm::ParserBase::parseBlockSignature):
(JSC::Wasm::ParserBase::parseReftypeSignature):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyModule.cpp:
(JSC::JSWebAssemblyModule::destroy):

Canonical link: <a href="https://commits.webkit.org/285969@main">https://commits.webkit.org/285969@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/88d5d6392768742cb7e682385da2f25099d32774

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/74337 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/53766 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/27148 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/78717 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/25575 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/62899 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/1551 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/58434 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/16756 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/77404 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/48598 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/63941 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/38843 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/45585 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/21436 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/23908 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/67474 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/66991 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/21783 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/80234 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/73595 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/1654 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/952 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/66719 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/1802 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/63959 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66002 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/9943 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/8096 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/95376 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11477 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/1618 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20946 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/1647 "Built successfully") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/129/builds/1635 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/1654 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

justinmichaud reviewed (approved):
r=me

There are so many places in the wasm runtime that make me worried about UAF. Callees (and the associated data held for refs, tables and imports) are another example. I would love to be able to clean up all of these places to make the ownership more explicit and predictable.

Constellation committed:
[JSC] Wasm GC BlockSignature should hold underlying TypeDefinition
https://bugs.webkit.org/show_bug.cgi?id=282363
rdar://138177632

Reviewed by Justin Michaud.

Wasm GC BlockSignature is not holding adhocly created TypeDefinition as
RefPtr. That's completely wrong: not holding TypeDefinition is allowed
because it is generated in wasm types in the section, and held by
ModuleInformation. If it is adhocly generated, it must be retained.

This patch fixes it by changing BlockSignature to keep underlying TypeDefinition
via RefPtr.

* JSTests/wasm/stress/block-signature-create-adhoc-type-definition.js: Added.
* Source/JavaScriptCore/wasm/WasmBBQJIT.cpp:
(JSC::Wasm::BBQJITImpl::ControlData::branchTargetArity const):
(JSC::Wasm::BBQJITImpl::ControlData::branchTargetType const):
(JSC::Wasm::BBQJITImpl::ControlData::argumentType const):
(JSC::Wasm::BBQJITImpl::BBQJIT::addBlock):
(JSC::Wasm::BBQJITImpl::BBQJIT::addLoop):
(JSC::Wasm::BBQJITImpl::BBQJIT::addIf):
(JSC::Wasm::BBQJITImpl::BBQJIT::addElse):
(JSC::Wasm::BBQJITImpl::BBQJIT::addElseToUnreachable):
(JSC::Wasm::BBQJITImpl::BBQJIT::addTry):
(JSC::Wasm::BBQJITImpl::BBQJIT::addTryTable):
(JSC::Wasm::BBQJITImpl::BBQJIT::addReturn):
(JSC::Wasm::BBQJITImpl::BBQJIT::addEndToUnreachable):
* Source/JavaScriptCore/wasm/WasmBBQJIT64.cpp:
(JSC::Wasm::BBQJITImpl::BBQJIT::addFusedIfCompare):
* Source/JavaScriptCore/wasm/WasmFormat.h:
* Source/JavaScriptCore/wasm/WasmFunctionParser.h:
(JSC::Wasm::splitStack):
(JSC::Wasm::FunctionParser<Context>::parseBlockSignatureAndNotifySIMDUseIfNeeded):
(JSC::Wasm::FunctionParser<Context>::parseBody):
(JSC::Wasm::FunctionParser<Context>::binaryCompareCase):
(JSC::Wasm::FunctionParser<Context>::unaryCompareCase):
(JSC::Wasm::FunctionParser<Context>::checkBranchTarget):
(JSC::Wasm::FunctionParser<Context>::unify):
(JSC::Wasm::FunctionParser<Context>::parseNestedBlocksEagerly):
(JSC::Wasm::FunctionParser<Context>::parseExpression):
* Source/JavaScriptCore/wasm/WasmIPIntGenerator.cpp:
(JSC::Wasm::IPIntControlType::branchTargetType const):
(JSC::Wasm::IPIntControlType::branchTargetArity const):
(JSC::Wasm::IPIntGenerator::addElseToUnreachable):
(JSC::Wasm::IPIntGenerator::addEndToUnreachable):
(JSC::Wasm::IPIntGenerator::endTopLevel):
* Source/JavaScriptCore/wasm/WasmLLIntGenerator.cpp:
(JSC::Wasm::LLIntGenerator::ControlType::loop):
(JSC::Wasm::LLIntGenerator::ControlType::block):
(JSC::Wasm::LLIntGenerator::ControlType::if_):
(JSC::Wasm::LLIntGenerator::ControlType::createTry):
(JSC::Wasm::LLIntGenerator::ControlType::createTryTable):
(JSC::Wasm::LLIntGenerator::ControlType::branchTargetArity const):
(JSC::Wasm::LLIntGenerator::ControlType::branchTargetType const):
(JSC::Wasm::LLIntGenerator::addElseToUnreachable):
(JSC::Wasm::LLIntGenerator::addReturn):
(JSC::Wasm::LLIntGenerator::addEndToUnreachable):
(JSC::Wasm::LLIntGenerator::endTopLevel):
* Source/JavaScriptCore/wasm/WasmModuleInformation.cpp:
(JSC::Wasm::ModuleInformation::operator delete):
* Source/JavaScriptCore/wasm/WasmModuleInformation.h:
(JSC::Wasm::ModuleInformation::create): Deleted.
(JSC::Wasm::ModuleInformation::functionIndexSpaceSize const): Deleted.
(JSC::Wasm::ModuleInformation::isImportedFunctionFromFunctionIndexSpace const): Deleted.
(JSC::Wasm::ModuleInformation::typeIndexFromFunctionIndexSpace const): Deleted.
(JSC::Wasm::ModuleInformation::exceptionIndexSpaceSize const): Deleted.
(JSC::Wasm::ModuleInformation::isImportedExceptionFromExceptionIndexSpace const): Deleted.
(JSC::Wasm::ModuleInformation::typeIndexFromExceptionIndexSpace const): Deleted.
(JSC::Wasm::ModuleInformation::importFunctionCount const): Deleted.
(JSC::Wasm::ModuleInformation::internalFunctionCount const): Deleted.
(JSC::Wasm::ModuleInformation::importExceptionCount const): Deleted.
(JSC::Wasm::ModuleInformation::internalExceptionCount const): Deleted.
(JSC::Wasm::ModuleInformation::toCodeIndex const): Deleted.
(JSC::Wasm::ModuleInformation::toSpaceIndex const): Deleted.
(JSC::Wasm::ModuleInformation::memoryCount const): Deleted.
(JSC::Wasm::ModuleInformation::tableCount const): Deleted.
(JSC::Wasm::ModuleInformation::elementCount const): Deleted.
(JSC::Wasm::ModuleInformation::globalCount const): Deleted.
(JSC::Wasm::ModuleInformation::dataSegmentsCount const): Deleted.
(JSC::Wasm::ModuleInformation::table const): Deleted.
(JSC::Wasm::ModuleInformation::global const): Deleted.
(JSC::Wasm::ModuleInformation::initializeFunctionTrackers const): Deleted.
(JSC::Wasm::ModuleInformation::referencedFunctions const): Deleted.
(JSC::Wasm::ModuleInformation::hasReferencedFunction const): Deleted.
(JSC::Wasm::ModuleInformation::addReferencedFunction const): Deleted.
(JSC::Wasm::ModuleInformation::isDeclaredFunction const): Deleted.
(JSC::Wasm::ModuleInformation::addDeclaredFunction): Deleted.
(JSC::Wasm::ModuleInformation::isDeclaredException const): Deleted.
(JSC::Wasm::ModuleInformation::addDeclaredException): Deleted.
(JSC::Wasm::ModuleInformation::functionWasmSizeImportSpace const): Deleted.
(JSC::Wasm::ModuleInformation::functionWasmSize const): Deleted.
(JSC::Wasm::ModuleInformation::usesSIMDImportSpace const): Deleted.
(JSC::Wasm::ModuleInformation::usesSIMD const): Deleted.
(JSC::Wasm::ModuleInformation::markUsesSIMD): Deleted.
(JSC::Wasm::ModuleInformation::usesExceptions const): Deleted.
(JSC::Wasm::ModuleInformation::markUsesExceptions): Deleted.
(JSC::Wasm::ModuleInformation::usesAtomics const): Deleted.
(JSC::Wasm::ModuleInformation::markUsesAtomics): Deleted.
(JSC::Wasm::ModuleInformation::doneSeeingFunction): Deleted.
(JSC::Wasm::ModuleInformation::typeCount const): Deleted.
(JSC::Wasm::ModuleInformation::hasMemoryImport const): Deleted.
(JSC::Wasm::ModuleInformation::getBranchHint const): Deleted.
(JSC::Wasm::ModuleInformation::clobberingTailCalls const): Deleted.
(JSC::Wasm::ModuleInformation::callCanClobberInstance const): Deleted.
(JSC::Wasm::ModuleInformation::addClobberingTailCall): Deleted.
* Source/JavaScriptCore/wasm/WasmOMGIRGenerator.cpp:
(JSC::Wasm::OMGIRGenerator::ControlData::ControlData):
(JSC::Wasm::OMGIRGenerator::ControlData::hasNonVoidresult const):
(JSC::Wasm::OMGIRGenerator::ControlData::branchTargetArity const):
(JSC::Wasm::OMGIRGenerator::ControlData::branchTargetType const):
(JSC::Wasm::OMGIRGenerator::addLoop):
(JSC::Wasm::OMGIRGenerator::addTopLevel):
(JSC::Wasm::OMGIRGenerator::addBlock):
(JSC::Wasm::OMGIRGenerator::addElseToUnreachable):
(JSC::Wasm::OMGIRGenerator::addEndToUnreachable):
* Source/JavaScriptCore/wasm/WasmParser.h:
(JSC::Wasm::ParserBase::parseBlockSignature):
(JSC::Wasm::ParserBase::parseReftypeSignature):
* Source/JavaScriptCore/wasm/js/JSWebAssemblyModule.cpp:
(JSC::JSWebAssemblyModule::destroy):

Canonical link: https://commits.webkit.org/285969@main
